projects/batfish/src/main/java/org/batfish/job/BatfishJobExecutor.java

```diff
@@ -41,105 +50,143 @@ public BatfishJobExecutor(
       Collections.shuffle(jobs);
     }
-    List<Future<JobResultT>> futures = new ArrayList<>();
+
     for (JobT job : jobs) {
-      Future<JobResultT> future = pool.submit(job);
-      futures.add(future);
+      completionService.submit(job);
     }
+
+    initializeJobsStats(jobs);
     boolean processingError = false;
-    int finishedJobs = 0;
-    int totalJobs = jobs.size();
-    AtomicInteger completed = Driver.newBatch(_settings, _description, totalJobs);
-    double finishedPercent;
     List<BatfishException> failureCauses = new ArrayList<>();
-    while (!futures.isEmpty()) {
-      List<Future<JobResultT>> currentFutures = new ArrayList<>();
-      currentFutures.addAll(futures);
-      for (Future<JobResultT> future : currentFutures) {
-        if (future.isDone()) {
-          futures.remove(future);
-          finishedJobs++;
-          completed.incrementAndGet();
-          finishedPercent = 100 * ((double) finishedJobs) / totalJobs;
-          JobResultT result = null;
-          try {
-            result = future.get();
-          } catch (InterruptedException | ExecutionException e) {
-            throw new BatfishException("Error executing job", e);
-          }
-          String time = CommonUtil.getTime(result.getElapsedTime());
-          Throwable failureCause = result.getFailureCause();
-          if (failureCause == null) {
-            result.applyTo(output, _logger, answerElement);
-            _logger.infof(
-                "Job terminated successfully with result: %s after elapsed time: %s - %d/%d "
-                    + "(%.1f%%) complete\n",
-                result.toString(), time, finishedJobs, totalJobs, finishedPercent);
-          } else {
-            String failureMessage =
-                "Failure running job after elapsed time: "
-                    + time
-                    + "\n-----BEGIN JOB LOG-----\n"
-                    + result
-                        .getHistory()
-                        .toString(BatfishLogger.getLogLevel(_settings.getLogLevel()))
-                    + "\n-----END JOB LOG-----";
-            BatfishException bfc = new BatfishException(failureMessage, failureCause);
-            if (_settings.getExitOnFirstError()) {
-              result.appendHistory(_logger);
-              throw bfc;
-            } else {
-              processingError = true;
-              result.appendHistory(_logger);
-              _logger.error(failureMessage + ":\n\t" + ExceptionUtils.getStackTrace(failureCause));
-              failureCauses.add(bfc);
-              if (!_haltOnProcessingError) {
-                result.applyTo(output, _logger, answerElement);
-              }
-            }
-          }
-        } else {
-          continue;
-        }
-      }
-      if (!futures.isEmpty()) {
+
+    for (int i = 0; i < jobs.size(); i++) {
+      try {
+        // getting the next completed task
+        Future<JobResultT> currentDoneFuture = completionService.take();
```

Comment:
The code inside a try should be minimal. UpdateJobStats, result getting, etc., should be outside the `try` clause.

Honestly, I'd probably just

```java
try {
    for (...) {
       result = completionService.take().get();
       ...
    }
} catch (...) {
} finally {
    pool.shutdownNow();
}
```

and handle the various exceptions acceptably.

* InterruptedException: this means the task didn't finish.
* ExecutionException: the task failed.

Reply:
Updatejobstats, resultgetting/processing has to be done in each iteration so this approach will need them inside the try catch, I can put the try catch inside the for... loop but that will need the pool.shutdown() method to be called in all catch blocks and also after the looping is done

Reply:
done, but not sure if this is the best way to do it